File: typescript/packages/mechanisms/evm/test/unit/batch-settlement/submit.test.ts
import { describe, it, expect } from "vitest";
import { privateKeyToAccount } from "viem/accounts";
import {
  assertDirectAuthorizerSubmitter,
  shouldRelaySubmit,
} from "../../../src/batch-settlement/facilitator/submit";
import type { FacilitatorEvmSigner } from "../../../src/signer";

function buildSubmitter(address: `0x${string}`): FacilitatorEvmSigner {
  return {
    getAddresses: () => [address],
    readContract: async () => undefined,
    verifyTypedData: async () => true,
    writeContract: async () => "0x" + "ab".repeat(32),
    sendTransaction: async () => "0x" + "cd".repeat(32),
    waitForTransactionReceipt: async () => ({ status: "success" }),
    getCode: async () => "0x6080604052",
  };
}

describe("batch-settlement submit helpers", () => {
  it("shouldRelaySubmit relays when an authorizer signature is already present", () => {
    expect(shouldRelaySubmit("direct", true)).toBe(true);
    expect(shouldRelaySubmit(undefined, true)).toBe(true);
  });

  it("shouldRelaySubmit uses direct dispatch only when configured and unsigned", () => {
    expect(shouldRelaySubmit("direct", false)).toBe(false);
    expect(shouldRelaySubmit("relay", false)).toBe(true);
    expect(shouldRelaySubmit(undefined, false)).toBe(true);
  });

  it("assertDirectAuthorizerSubmitter is a no-op outside direct mode", () => {
    expect(() => assertDirectAuthorizerSubmitter("relay", undefined, undefined)).not.toThrow();
  });

  it("assertDirectAuthorizerSubmitter requires a dedicated authorizer signer and submitter", () => {
    const authorizer = privateKeyToAccount(
      "0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d",
    );
    expect(() =>
      assertDirectAuthorizerSubmitter("direct", undefined, buildSubmitter(authorizer.address)),
    ).toThrow('submitMode "direct" requires authorizerSigner');
    expect(() =>
      assertDirectAuthorizerSubmitter(
        "direct",
        { address: authorizer.address, signTypedData: async () => "0x" + "11".repeat(65) },
        undefined,
      ),
    ).toThrow('submitMode "direct" requires authorizerSubmitter');
  });

  it("assertDirectAuthorizerSubmitter rejects a submitter that is not exactly the authorizer", () => {
    const authorizer = privateKeyToAccount(
      "0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d",
    );
    const other = privateKeyToAccount(
      "0x5de4111afa1a4b94908f83103eb1f1706367c2e68ca870fc3fb9a804cdab365a",
    );
    expect(() =>
      assertDirectAuthorizerSubmitter(
        "direct",
        { address: authorizer.address, signTypedData: async () => "0x" + "11".repeat(65) },
        buildSubmitter(other.address),
      ),
    ).toThrow("authorizerSubmitter.getAddresses() must be exactly [authorizerSigner.address]");
  });

  it("assertDirectAuthorizerSubmitter accepts a single-address submitter matching the authorizer", () => {
    const authorizer = privateKeyToAccount(
      "0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d",
    );
    expect(() =>
      assertDirectAuthorizerSubmitter(
        "direct",
        { address: authorizer.address, signTypedData: async () => "0x" + "11".repeat(65) },
        buildSubmitter(authorizer.address),
      ),
    ).not.toThrow();
  });
});

File: go/mechanisms/evm/batch-settlement/facilitator/submit_test.go
package facilitator

import (
	"strings"
	"testing"
)

func TestShouldRelaySubmit_RelaysWhenAuthorizerSignaturePresent(t *testing.T) {
	if !ShouldRelaySubmit(SubmitModeDirect, true) {
		t.Fatal("direct + signature should relay")
	}
	if !ShouldRelaySubmit("", true) {
		t.Fatal("omitted mode + signature should relay")
	}
}

func TestShouldRelaySubmit_DirectOnlyWhenConfiguredAndUnsigned(t *testing.T) {
	if ShouldRelaySubmit(SubmitModeDirect, false) {
		t.Fatal("direct + unsigned should not relay")
	}
	if !ShouldRelaySubmit(SubmitModeRelay, false) {
		t.Fatal("relay + unsigned should relay")
	}
	if !ShouldRelaySubmit("", false) {
		t.Fatal("omitted mode + unsigned should relay")
	}
}

func TestAssertDirectAuthorizerSubmitter_NoOpOutsideDirect(t *testing.T) {
	if err := AssertDirectAuthorizerSubmitter(SubmitModeRelay, nil, nil); err != nil {
		t.Fatalf("relay should be a no-op: %v", err)
	}
}

func TestAssertDirectAuthorizerSubmitter_RequiresSignerAndSubmitter(t *testing.T) {
	submitter := &fakeFacilitatorSigner{addresses: []string{"0x70997970C51812dc3A010C7d01b50e0d17dc79C8"}}
	err := AssertDirectAuthorizerSubmitter(SubmitModeDirect, nil, submitter)
	if err == nil || !strings.Contains(err.Error(), `submitMode "direct" requires authorizerSigner`) {
		t.Fatalf("got %v", err)
	}
	authorizer := &fakeAuthorizerSigner{addr: "0x70997970C51812dc3A010C7d01b50e0d17dc79C8"}
	err = AssertDirectAuthorizerSubmitter(SubmitModeDirect, authorizer, nil)
	if err == nil || !strings.Contains(err.Error(), `submitMode "direct" requires authorizerSubmitter`) {
		t.Fatalf("got %v", err)
	}
}

func TestAssertDirectAuthorizerSubmitter_RejectsMismatchedSubmitter(t *testing.T) {
	authorizer := &fakeAuthorizerSigner{addr: "0x70997970C51812dc3A010C7d01b50e0d17dc79C8"}
	submitter := &fakeFacilitatorSigner{addresses: []string{"0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC"}}
	err := AssertDirectAuthorizerSubmitter(SubmitModeDirect, authorizer, submitter)
	if err == nil || !strings.Contains(err.Error(), "authorizerSubmitter.getAddresses() must be exactly [authorizerSigner.address]") {
		t.Fatalf("got %v", err)
	}
}

func TestAssertDirectAuthorizerSubmitter_AcceptsMatchingSingleAddress(t *testing.T) {
	addr := "0x70997970C51812dc3A010C7d01b50e0d17dc79C8"
	authorizer := &fakeAuthorizerSigner{addr: addr}
	submitter := &fakeFacilitatorSigner{addresses: []string{addr}}
	if err := AssertDirectAuthorizerSubmitter(SubmitModeDirect, authorizer, submitter); err != nil {
		t.Fatal(err)
	}
}
